lib/srv/app/server.go

```diff
@@ -640,39 +749,48 @@ func (s *Server) handleHTTPApp(ctx context.Context, conn net.Conn) error {
 
 // ServeHTTP will forward the *http.Request to the target application.
 func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	if err := s.serveHTTP(w, r); err != nil {
+	// See if the initial auth failed. If it didn't, serve the HTTP regularly, which
+	// will include subsequent auth attempts to prevent race-type conditions.
+	err := s.getAndDeleteConnAuth(r.Context().Value(connContextKey).(net.Conn))
+	if err == nil {
+		err = s.serveHTTP(w, r)
+	}
```

Comment:
So, basically if I just return an error during the initial auth, Teleport responds with a 500 error for HTTP apps rather than the expected 403. This answers the above issue about why it handles these errors differently -- I figured TCP apps should error out quickly and HTTP apps should push back the 403 during this issue.